presto-raptor/src/test/java/com/facebook/presto/raptor/storage/TestOrcStorageManager.java

```diff
@@ -371,6 +377,136 @@ public void testRewriter()
+        Collection<Slice> fragments = deltaDelete(rowsToDelete, blockBuilder.build(), false);
+
+        assertEquals(ImmutableList.of(), fragments);
+        List<RecordedShard> recordedShards = shardRecorder.getShards();
+        assertEquals(recordedShards.size(), 1);
+    }
+
+    @Test
+    public void testRewriterWithDeltaDeleteAll()
+    {
+        // delete every row
+        BitSet rowsToDelete = new BitSet();
+        rowsToDelete.set(0);
+        rowsToDelete.set(1);
+        rowsToDelete.set(2);
+        LongArrayBlockBuilder blockBuilder = new LongArrayBlockBuilder(null, 1);
+        blockBuilder.writeLong(0).writeLong(1).writeLong(2);
+        Collection<Slice> fragments = deltaDelete(rowsToDelete, blockBuilder.build(), false);
+
+        Slice shardDelta = Iterables.getOnlyElement(fragments);
+        ShardDeleteDelta shardDeltas = jsonCodec(ShardDeleteDelta.class).fromJson(shardDelta.getBytes());
+        assertEquals(shardDeltas.getNewDeltaDeleteShard(), Optional.empty());
+
+        // verify recorded shard
+        List<RecordedShard> recordedShards = shardRecorder.getShards();
+        assertEquals(recordedShards.size(), 1);
+    }
+
+    @Test
+    public void testRewriterWithDeltaDeleteMerge()
+    {
+        // delete every row
+        BitSet rowsToDelete = new BitSet();
+        rowsToDelete.set(0);
+        rowsToDelete.set(1);
+        LongArrayBlockBuilder blockBuilder = new LongArrayBlockBuilder(null, 1);
+        blockBuilder.writeLong(0).writeLong(1);
+
+        Collection<Slice> fragments = deltaDelete(rowsToDelete, blockBuilder.build(), true);
+
+        Slice shardDelta = Iterables.getOnlyElement(fragments);
+        ShardDeleteDelta shardDeltas = jsonCodec(ShardDeleteDelta.class).fromJson(shardDelta.getBytes());
+        assertEquals(shardDeltas.getNewDeltaDeleteShard(), Optional.empty());
+
+        // verify recorded shard
+        List<RecordedShard> recordedShards = shardRecorder.getShards();
+        assertEquals(recordedShards.size(), 2);
+    }
+
+    private Collection<Slice> deltaDelete(BitSet rowsToDelete, Block blockToDelete, boolean oldDeltaDeleteExist)
+    {
+        OrcStorageManager manager = createOrcStorageManager();
+
+        long transactionId = TRANSACTION_ID;
+        List<Long> columnIds = ImmutableList.of(3L, 7L);
+        List<Type> columnTypes = ImmutableList.of(BIGINT, createVarcharType(10));
+
+        // create file with 2 rows
+        StoragePageSink sink = createStoragePageSink(manager, columnIds, columnTypes);
+        List<Page> pages = rowPagesBuilder(columnTypes)
+                .row(123L, "hello")
+                .row(456L, "bye")
+                .row(456L, "test")
+                .build();
+        sink.appendPages(pages);
+        List<ShardInfo> shards = getFutureValue(sink.commit());
+        assertEquals(shardRecorder.getShards().size(), 1);
+
+        List<ShardInfo> deltaDeleteShards = null;
+        if (oldDeltaDeleteExist) {
```

Comment:
I guess you don't need to manually create a delta? You can just run delta delete twice and the second one will encounter an existing delta.